File: momentra/momentra/Shell/Personal/Relationships/Create/PersonalRelationshipsDerived.swift
import Foundation

/// Relationships Bond Index + axis scores from pulse + widgetPayload (Figma 505:11793).
enum PersonalRelationshipsDerived {
    struct BondAxes {
        let trust: String
        let care: String
        let support: String
        let presence: String
    }

    static func payloadNumber(_ payload: [String: AnyDecodable]?, key: String) -> Double? {
        guard let raw = payload?[key]?.value else { return nil }
        if let n = raw as? Double { return n }
        if let n = raw as? Int { return Double(n) }
        if let s = raw as? String { return PersonalLifeOpsDerived.scoreNumber(s) }
        if let n = raw as? NSNumber { return n.doubleValue }
        return nil
    }

    static func displayFromPayloadOrFallback(
        payload: [String: AnyDecodable]?,
        payloadKey: String,
        fallback: String?
    ) -> String {
        if let n = payloadNumber(payload, key: payloadKey) {
            return String(Int(n.rounded()))
        }
        return PersonalLifeOpsDerived.displayScore(fallback)
    }

    static func bondIndex(pulse: APIClient.PersonalPulsePayload?) -> Int? {
        if let w = PersonalLifeOpsDerived.scoreNumber(pulse?.wellbeingScore) { return Int(w.rounded()) }
        if let r = PersonalLifeOpsDerived.scoreNumber(pulse?.rhythmScore) { return Int(r.rounded()) }
        let keys = ["bondIndex", "trustScore", "careScore", "supportScore", "presenceScore"]
        let nums = keys.compactMap { payloadNumber(pulse?.widgetPayload, key: $0) }
        guard !nums.isEmpty else { return nil }
        return Int((nums.reduce(0, +) / Double(nums.count)).rounded())
    }

    static func bondIndexDisplay(pulse: APIClient.PersonalPulsePayload?) -> String {
        bondIndex(pulse: pulse).map(String.init) ?? "—"
    }

    static func bondAxes(pulse: APIClient.PersonalPulsePayload?) -> BondAxes {
        let payload = pulse?.widgetPayload
        return BondAxes(
            trust: displayFromPayloadOrFallback(payload: payload, payloadKey: "trustScore", fallback: nil),
            care: displayFromPayloadOrFallback(payload: payload, payloadKey: "careScore", fallback: nil),
            support: displayFromPayloadOrFallback(payload: payload, payloadKey: "supportScore", fallback: nil),
            presence: displayFromPayloadOrFallback(payload: payload, payloadKey: "presenceScore", fallback: nil)
        )
    }

    static func bondSubtitle(pulse: APIClient.PersonalPulsePayload?) -> String {
        bondIndex(pulse: pulse) != nil ? "Bond signal present" : "Awaiting bond signals"
    }

    static func spendPairs(pulse: APIClient.PersonalPulsePayload?) -> [(String, String)] {
        guard let map = pulse?.widgetPayload?["spendByCurrency"]?.value as? [String: Any] else { return [] }
        return map.compactMap { key, value in
            let v = "\(value)"
            guard !key.isEmpty, !v.isEmpty else { return nil }
            return (key, v)
        }
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/personal/relationships/create/PersonalRelationshipsDerived.kt
package com.example.momentra.ui.shell.personal.relationships.create

import com.example.momentra.data.api.PersonalPulseDto
import kotlin.math.roundToInt
import com.example.momentra.ui.shell.personal.lifeops.create.PersonalLifeOpsDerived

/** Relationships Bond Index from pulse widgetPayload (PX-3 precision refresh — no fixed heuristics). */
object PersonalRelationshipsDerived {
    data class BondAxes(
        val trust: String,
        val care: String,
        val support: String,
        val presence: String,
    )

    fun payloadNumber(payload: Map<String, Any>?, key: String): Double? {
        val raw = payload?.get(key) ?: return null
        return when (raw) {
            is Number -> raw.toDouble()
            is String -> PersonalLifeOpsDerived.scoreNumber(raw)
            else -> null
        }
    }

    fun displayFromPayloadOrFallback(
        payload: Map<String, Any>?,
        payloadKey: String,
        fallback: String?,
    ): String {
        val fromPayload = payloadNumber(payload, payloadKey)
        if (fromPayload != null) return fromPayload.roundToInt().toString()
        return PersonalLifeOpsDerived.displayScore(fallback)
    }

    fun bondIndex(pulse: PersonalPulseDto?): Int? {
        PersonalLifeOpsDerived.scoreNumber(pulse?.wellbeingScore)?.roundToInt()?.let { return it }
        PersonalLifeOpsDerived.scoreNumber(pulse?.rhythmScore)?.roundToInt()?.let { return it }
        val payload = pulse?.widgetPayload
        val axes = listOf("bondIndex", "trustScore", "careScore", "supportScore", "presenceScore")
            .mapNotNull { payloadNumber(payload, it) }
        if (axes.isEmpty()) return null
        return axes.average().roundToInt()
    }

    fun bondIndexDisplay(pulse: PersonalPulseDto?): String =
        bondIndex(pulse)?.toString() ?: "—"

    fun bondAxes(pulse: PersonalPulseDto?): BondAxes {
        val payload = pulse?.widgetPayload
        return BondAxes(
            trust = displayFromPayloadOrFallback(payload, "trustScore", null),
            care = displayFromPayloadOrFallback(payload, "careScore", null),
            support = displayFromPayloadOrFallback(payload, "supportScore", null),
            presence = displayFromPayloadOrFallback(payload, "presenceScore", null),
        )
    }

    fun bondSubtitle(pulse: PersonalPulseDto?): String {
        val score = bondIndex(pulse)
        return if (score != null) "Bond signal present" else "Awaiting bond signals"
    }

    fun spendPairs(pulse: PersonalPulseDto?): List<Pair<String, String>> =
        (pulse?.widgetPayload?.get("spendByCurrency") as? Map<*, *>)
            ?.entries
            ?.mapNotNull { (k, v) ->
                val key = k?.toString()?.takeIf { it.isNotBlank() } ?: return@mapNotNull null
                val value = v?.toString()?.takeIf { it.isNotBlank() } ?: return@mapNotNull null
                key to value
            }
            .orEmpty()
}
